### PR TITLE
Implement loss function for multi label classification tasks

### PR DESCRIPTION
**Pull request by treo**
This implements the loss function from this paper:

Min-Ling Zhang and Zhi-Hua Zhou, "Multilabel Neural Networks with Applications to Functional Genomics and Text Categorization," in IEEE Transactions on Knowledge and Data Engineering, vol. 18, no. 10, pp. 1338-1351, Oct. 2006.

It is useful as a loss function for multi label classification tasks.

### CODE REVIEW

**Comment by saudet** [comment]
@sshepel Did you set some commands to force Jenkins to rebuild the branch without having to push new commits?

**Comment by sshepel** [comment]
> @sshepel Did you set some commands to force Jenkins to rebuild the branch without having to push new commits?
There is a button in Jenkins UI to do that, I have already provided @treo with all required info, no additional forcing logic where added.

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+            final INDArray pairwiseSub = Transforms.exp(operandA.sub(operandB));
+
+            final INDArray selection = locPositive.transpose().mmul(locNegative);
+
+            final INDArray classificationDifferences = pairwiseSub.muli(selection).divi(locNormFactor);
+
+            if (scoreOutput != null) {
+                if (mask != null) {
+                    final INDArray perLabel = classificationDifferences.sum(0);
+                    LossUtil.applyMask(perLabel, mask);
```
Comment (AlexDBlack) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java]: This line looks incorrect at first glance... seems like you are applying the full (all examples) mask on a single row?

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        }
+        final INDArray postOutput = activationFn.getActivation(preOutput.dup(), true);
+
+        final INDArray positive = labels;
+        final INDArray negative = labels.eq(0.0);
+        final INDArray normFactor = negative.sum(1).muli(positive.sum(1));
+
+
+        int examples = positive.size(0);
+        for (int i = 0; i < examples; i++) {
```
Comment (AlexDBlack) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java]: I haven't checked the math - but I assume there's no way to do this without the loop?

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+            final INDArray locCfn = postOutput.getRow(i);
+            final int[] shape = locCfn.shape();
+
+            final INDArray locPositive = positive.getRow(i);
+            final INDArray locNegative = negative.getRow(i);
+            final INDArray locNormFactor = normFactor.getScalar(i);
```
Comment (AlexDBlack) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java]: Why not getDouble? Scalar in INDArray seems to provide (minor) overhead with no benefit?

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+                    final INDArray perLabel = classificationDifferences.sum(0);
+                    LossUtil.applyMask(perLabel, mask);
+                    perLabel.sum(scoreOutput.getRow(i), 0);
+                } else {
+                    classificationDifferences.sum(scoreOutput.getRow(i), 0, 1);
+                }
+            }
+
+            if (gradientOutput != null) {
+                gradientOutput.getRow(i).assign(classificationDifferences.sum(0).add(classificationDifferences.sum(1).transposei().negi()));
```
Comment (AlexDBlack) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java]: Minor optimization: sum(0).add*i*

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    }
+
+    @Override
+    public SDVariable[] outputVariables(String baseName) {
+        return new SDVariable[0];
+    }
+
+    @Override
+    public List<SDVariable> doDiff(List<SDVariable> f1) {
+        return null;
```
Comment (AlexDBlack) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java]: I'd rather an UnsupportedOperationException

Comment (treo) [reply nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java]: I've copied that over from the other loss functions. Not quite sure what exactly is expected here, so I've left it the way that it is with the others.

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java (hunk 3 above)
Comment (treo) [reply nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java]: I was trying to keep the data movement off-heap->heap->off-heap to a minimum. But now that you've asked I actually looked at what getScalar does, and seen that it does just call getDouble internally... So I'll change that.

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java (hunk 2 above)
Comment (treo) [reply nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java]: I've failed to find a way that works without the loop. Every method that I could come up with, that skips the loop also calculates A LOT of unnecessary results. The speed hit isn't as bad as it may look like. It is at about 85% to 90% of BinaryXENT (e.g. 130 batches per second vs 150 batches per second).

If you have an idea how it could be done any faster, or any more vectorized, I'd love to hear it :)

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java (hunk 1 above)
Comment (treo) [reply nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMultiLabel.java]: You've got a point here, I've been moving around the masking stuff to find a proper place for it, and missed calling getRow here.
